File: ios/App/App/pont/StockageToken.swift
// Cree le 25/08/2026, Bourama : Lot 3B Partie 3 mobile (fusion Capacitor), iOS.
// Equivalent exact de StockageToken.kt (Android, meme lot) : Keychain plutot
// que EncryptedSharedPreferences, meme role.
//
// Choix d'architecture (identique Android) : contrairement au socle natif
// clovis-mobile (SupabaseAuthClient.swift, sa PROPRE auth Supabase native),
// ce plugin n'authentifie plus rien lui-meme, l'utilisateur est deja
// connecte cote WEB (clovis-frontend, dans la WKWebView Capacitor). Le pont
// recoit le token d'acces directement du JS et le garde dans le Keychain
// (chiffre par le systeme), necessaire pour que les notifications push
// puissent appeler clovis-backend meme app fermee.
import Foundation
import Security

enum StockageToken {

    private static let service = "ai.clovis.pont"
    private static let compte = "supabase_access_token"
    // Ajoute le 12/09/2026 : voir StockageToken.kt (Android, meme date) pour
    // le contexte complet du correctif.
    private static let compteTokenPush = "dernier_token_push_apns"

    private static func enregistrer(_ valeur: String, compte: String) {
        let donnees = Data(valeur.utf8)
        let requeteSuppression: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: compte
        ]
        SecItemDelete(requeteSuppression as CFDictionary)

        let requeteAjout: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: compte,
            kSecValueData as String: donnees,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock
        ]
        SecItemAdd(requeteAjout as CFDictionary, nil)
    }

    private static func lire(compte: String) -> String? {
        let requete: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: compte,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var resultat: AnyObject?
        let statut = SecItemCopyMatching(requete as CFDictionary, &resultat)
        guard statut == errSecSuccess, let donnees = resultat as? Data else { return nil }
        return String(data: donnees, encoding: .utf8)
    }

    private static func effacer(compte: String) {
        let requete: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: compte
        ]
        SecItemDelete(requete as CFDictionary)
    }

    static func enregistrer(_ token: String) {
        enregistrer(token, compte: compte)
    }

    static func lire() -> String? {
        lire(compte: compte)
    }

    static func effacer() {
        effacer(compte: compte)
    }

    static func enregistrerTokenPush(_ tokenPush: String) {
        enregistrer(tokenPush, compte: compteTokenPush)
    }

    static func lireTokenPush() -> String? {
        lire(compte: compteTokenPush)
    }
}

File: android/app/src/main/java/com/clovis/app/pont/StockageToken.kt
// Cree le 25/08/2026, Bourama : Lot 3B Partie 3 mobile (fusion Capacitor).
//
// Choix d'architecture : contrairement au socle natif clovis-mobile (qui
// avait sa PROPRE auth Supabase native, voir SupabaseAuthClient.kt), ce
// plugin n'authentifie plus rien lui-meme ; l'utilisateur est deja
// connecte cote WEB (clovis-frontend, dans la WebView Capacitor, via le
// SDK Supabase JS existant). Dupliquer une deuxieme session native aurait
// recree exactement le probleme de compte separe deja rencontre
// (connexion native qui echouait independamment de la session web).
//
// Le pont recoit donc le token d'acces directement du JS (voir
// PontNatifPlugin.enregistrerToken, appele par clovis-frontend apres
// connexion/rafraichissement Supabase) et le garde en local chiffre,
// necessaire pour que le service FCM puisse appeler clovis-backend meme
// quand l'app est reveillee en tache de fond, WebView fermee.
package com.clovis.app.pont

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

private const val FICHIER_PREFS = "clovis_pont_securise"
private const val CLE_TOKEN = "supabase_access_token"
// Ajoute le 12/09/2026 : dernier token push FCM obtenu de Google, garde en
// local independamment du succes de son envoi au serveur. onNewToken
// (ClovisFirebaseMessagingService) peut se declencher avant que le JS ait
// eu le temps d'appeler enregistrerToken (session Supabase pas encore
// prete) -- sans ce stockage, ce premier envoi echouait silencieusement
// (401, jamais verifie) et n'etait jamais retente, car Firebase ne
// redonne generalement ce token qu'une seule fois par installation.
private const val CLE_TOKEN_PUSH = "dernier_token_push_fcm"

object StockageToken {

    private fun prefs(context: Context) = EncryptedSharedPreferences.create(
        context,
        FICHIER_PREFS,
        MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )

    fun enregistrer(context: Context, token: String) {
        prefs(context).edit().putString(CLE_TOKEN, token).apply()
    }

    fun lire(context: Context): String? = prefs(context).getString(CLE_TOKEN, null)

    fun effacer(context: Context) {
        prefs(context).edit().remove(CLE_TOKEN).apply()
    }

    fun enregistrerTokenPush(context: Context, tokenPush: String) {
        prefs(context).edit().putString(CLE_TOKEN_PUSH, tokenPush).apply()
    }

    fun lireTokenPush(context: Context): String? = prefs(context).getString(CLE_TOKEN_PUSH, null)
}
